File: benchmarks/io/FileBufferedReadWrite.c
#ifdef _WIN32
#define _CRT_SECURE_NO_WARNINGS
#endif

#include <stdint.h>
#include <stdio.h>

static int64_t sum_bytes(const int8_t *values, int64_t count) {
    int64_t checksum = 0;
    for (int64_t index = 0; index < count; index += 1) {
        checksum += values[index];
    }

    return checksum;
}

int main(void) {
    static const int iterations = 64;
    static const int chunks = 33;
    const int8_t source[32] = {
        1, 2, 3, 4, 5, 6, 7, 8,
        9, 10, 11, 12, 13, 14, 15, 16,
        1, 2, 3, 4, 5, 6, 7, 8,
        9, 10, 11, 12, 13, 14, 15, 16
    };
    int8_t destination[32] = {0};
    int64_t checksum = 0;

    remove("experimental-buffered-rw.tmp");

    for (int iteration = 0; iteration < iterations; iteration += 1) {
        FILE *writer = fopen("experimental-buffered-rw.tmp", "wb");
        if (writer == NULL) {
            return 1;
        }

        for (int chunk = 0; chunk < 32; chunk += 1) {
            if (fwrite(source, 1, sizeof(source), writer) != sizeof(source)) {
                fclose(writer);
                return 1;
            }
        }

        if (fwrite(source, 1, sizeof(source), writer) != sizeof(source) ||
            fclose(writer) != 0) {
            return 1;
        }

        FILE *reader = fopen("experimental-buffered-rw.tmp", "rb");
        if (reader == NULL) {
            return 1;
        }

        if (fseek(reader, 0, SEEK_SET) != 0) {
            fclose(reader);
            return 1;
        }

        for (int chunk = 0; chunk < chunks; chunk += 1) {
            if (fread(destination, 1, sizeof(destination), reader) != sizeof(destination)) {
                fclose(reader);
                return 1;
            }

            checksum += sum_bytes(destination, 32);
        }

        fclose(reader);
        if (remove("experimental-buffered-rw.tmp") != 0) {
            return 1;
        }
    }

    return checksum == 574464 ? 0 : 1;
}

File: benchmarks/io/FileBufferedReadWrite.rs
use std::fs::{remove_file, File};
use std::io::{Read, Seek, SeekFrom, Write};

fn sum_bytes(values: &[i8]) -> i64 {
    let mut checksum = 0_i64;
    for value in values {
        checksum += *value as i64;
    }

    checksum
}

fn main() {
    const ITERATIONS: usize = 64;
    const CHUNKS: usize = 33;
    let source: [i8; 32] = [
        1, 2, 3, 4, 5, 6, 7, 8,
        9, 10, 11, 12, 13, 14, 15, 16,
        1, 2, 3, 4, 5, 6, 7, 8,
        9, 10, 11, 12, 13, 14, 15, 16,
    ];
    let mut destination = [0_u8; 32];
    let bytes: Vec<u8> = source.iter().map(|value| *value as u8).collect();
    let mut checksum = 0_i64;

    let _ = remove_file("experimental-buffered-rw.tmp");

    for _ in 0..ITERATIONS {
        {
            let mut writer = File::create("experimental-buffered-rw.tmp").expect("create");
            for _ in 0..32 {
                writer.write_all(&bytes).expect("write");
            }
            writer.write_all(&bytes).expect("write tail");
        }

        {
            let mut reader = File::open("experimental-buffered-rw.tmp").expect("open");
            reader.seek(SeekFrom::Start(0)).expect("seek");
            for _ in 0..CHUNKS {
                reader.read_exact(&mut destination).expect("read");
                let signed: Vec<i8> = destination.iter().map(|value| *value as i8).collect();
                checksum += sum_bytes(&signed);
            }
        }

        remove_file("experimental-buffered-rw.tmp").expect("delete");
    }

    if checksum != 574_464 {
        std::process::exit(1);
    }
}
